src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/HandshakeProtocol.cs

```diff
@@ -0,0 +1,136 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Buffers;
+using System.Collections;
+using System.IO;
+using System.Text;
+using Microsoft.AspNetCore.SignalR.Internal.Formatters;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace Microsoft.AspNetCore.SignalR.Internal.Protocol
+{
+    public static class HandshakeProtocol
+    {
+        private static readonly UTF8Encoding _utf8NoBom = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);
+
+        private const string ProtocolPropertyName = "protocol";
+        private const string ErrorPropertyName = "error";
+
+        public static void WriteRequestMessage(HandshakeRequestMessage requestMessage, Stream output)
+        {
+            using (var writer = CreateJsonTextWriter(output))
+            {
+                writer.WriteStartObject();
+                writer.WritePropertyName(ProtocolPropertyName);
+                writer.WriteValue(requestMessage.Protocol);
+                writer.WriteEndObject();
+            }
+
+            TextMessageFormatter.WriteRecordSeparator(output);
+        }
+
+        public static void WriteResponseMessage(HandshakeResponseMessage responseMessage, Stream output)
+        {
+            using (var writer = CreateJsonTextWriter(output))
+            {
+                writer.WriteStartObject();
+                if (!string.IsNullOrEmpty(responseMessage.Error))
+                {
+                    writer.WritePropertyName(ErrorPropertyName);
+                    writer.WriteValue(responseMessage.Error);
+                }
+                writer.WriteEndObject();
+            }
+
+            TextMessageFormatter.WriteRecordSeparator(output);
+        }
+
+        private static JsonTextWriter CreateJsonTextWriter(Stream output)
+        {
+            return new JsonTextWriter(new StreamWriter(output, _utf8NoBom, 1024, leaveOpen: true));
```

Comment:
nit: can we make it smaller than 1024?

Reply:
These streams are going to get a bit of an overhaul when we get to #1508 . I wouldn't optimize too heavily here until we know how that's going to land.

Reply:
1024 is the StreamWriter default. Lets leave it for now and come back when analyzing perf

Reply:
Yes, I want to get to these changes too but not as part of this PR.